src/vmm/src/devices/virtio/transport/mmio.rs

```diff
@@ -375,13 +376,56 @@ pub enum IrqType {
     Vring,
 }
 
+impl From<VirtioInterruptType> for IrqType {
+    fn from(interrupt_type: VirtioInterruptType) -> Self {
+        match interrupt_type {
+            VirtioInterruptType::Config => IrqType::Config,
+            VirtioInterruptType::Queue(_) => IrqType::Vring,
+        }
+    }
+}
+
 /// Helper struct that is responsible for triggering guest IRQs
 #[derive(Debug)]
 pub struct IrqTrigger {
     pub(crate) irq_status: Arc<AtomicU32>,
     pub(crate) irq_evt: EventFd,
 }
 
+impl VirtioInterrupt for IrqTrigger {
+    fn trigger(&self, interrupt_type: VirtioInterruptType) -> Result<(), std::io::Error> {
+        match interrupt_type {
+            VirtioInterruptType::Config => self.trigger_irq(IrqType::Config),
+            VirtioInterruptType::Queue(_) => self.trigger_irq(IrqType::Vring),
+        }
+    }
+
+    fn notifier(&self, _interrupt_type: VirtioInterruptType) -> Option<EventFd> {
+        self.irq_evt.try_clone().ok()
+    }
+
+    fn status(&self) -> Arc<AtomicU32> {
+        self.irq_status.clone()
+    }
+
+    #[cfg(test)]
+    fn has_pending_interrupt(&self, interrupt_type: VirtioInterruptType) -> bool {
+        if let Ok(num_irqs) = self.irq_evt.read() {
+            if num_irqs == 0 {
+                return false;
+            }
+
+            let irq_status = self.irq_status.load(Ordering::SeqCst);
+            return matches!(
+                (irq_status, interrupt_type.into()),
+                (VIRTIO_MMIO_INT_CONFIG, IrqType::Config) | (VIRTIO_MMIO_INT_VRING, IrqType::Vring)
+            );
+        }
+
+        false
+    }
```

Comment:
Can this just be a free-standing function `has_pending_interrupt<T: VirtioInterrupt>(&T, VirtioInterruptType) -> bool` somewhere inside a test module, so we avoid the `#[cfg(test)]`?

Reply:
hmmm. The problem is that for as long as we support both MMIO and PCIe we will not have an easy way to create an type independent implementation. WDYT?